compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala

```diff
@@ -928,6 +969,128 @@ abstract class GenJSCode[G <: Global with Singleton](val global: G)
       }
     }
 
+    // Static forwarders -------------------------------------------------------
+
+    /* This mimics the logic in BCodeHelpers.addForwarders and the code that
+     * calls it, except that we consider as collisions only methods with the
+     * same `MethodName` (including signature) and that are also
+     * `PublicStatic`. Anything else is not a collision as far the IR is
+     * concerned. We need this in some cases, for example the class `Integer`
+     * has a `def hashCode(): Int` and a `static def hashCode(Int): Int`. The
+     * JVM back-end considers them as colliding because they have the same
+     * name, but we must not.
+     */
+
+    /** Is the given Scala class, interface or module class a candidate for
+     *  static forwarders?
+     */
+    def isCandidateForForwarders(sym: Symbol): Boolean = {
+      // it must be a top level class (name contains no $s)
+      !settings.noForwarders && exitingPickler {
+        !sym.name.containsChar('$') && !sym.isNestedClass
+      }
+    }
+
+    /** Adds static forwarders to the members of a class or interface for
+     *  methods of its companion object.
+     *
+     *  This is only done if there exists a companion object and it is not a JS
+     *  type.
+     *
+     *  Precondition: `isCandidateForForwarders(sym)` is true
+     */
+    def addStaticForwardersForClassOrInterface(
+        existingMembers: List[js.MemberDef], sym: Symbol)(
+        implicit pos: Position): List[js.MemberDef] = {
+      val module = sym.companionModule
+      if (module == NoSymbol) {
+        existingMembers
+      } else {
+        val moduleClass = module.moduleClass
+        if (!isJSType(moduleClass))
+          addStaticFowardersFromModuleClass(existingMembers, moduleClass)
+        else
+          existingMembers
+      }
+    }
+
+    private lazy val dontUseExitingUncurryForForwarders =
+      scala.util.Properties.versionNumberString.startsWith("2.11.")
+
+    /** Adds static forwarders for the methods of a module class to a list of
+     *  existing members.
+     *
+     *  Precondition: `isCandidateForForwarders(moduleClass)` is true
+     */
+    def addStaticFowardersFromModuleClass(existingMembers: List[js.MemberDef],
+        moduleClass: Symbol)(
+        implicit pos: Position): List[js.MemberDef] = {
+
+      assert(moduleClass.isModuleClass, moduleClass)
+
+      val conflictingNames: Set[MethodName] = {
+        existingMembers.collect {
+          case js.MethodDef(flags, name, _, _, _, _)
+              if flags.namespace == js.MemberNamespace.PublicStatic =>
+            name.name
+        }.toSet
+      }
```

Comment:
So this should be `assert(!exisitngMembers.exists(_.flags.namespace == PublicStatic)`?

Reply:
I preferred using a nice `reporter.error` with a position, so that potential bug reports can be more useful. Also I'm doing it only when there is a conflict, since I can imagine a situation, in the future, where scalac would generate private static methods in a class for its own purposes, that would then be made public to be accessible from a different class, but in that case they would have a mangled name, which could not collide with the static forwarders (we do not generate forwarders for methods with mangled names).

Reply:
> I preferred using a nice reporter.error with a position, so that potential bug reports can be more useful. 

That's fair.

> Also I'm doing it only when there is a conflict [...]

Since this is not necessary right now, personally, I'd feel more comfortable with checking the (more rigid) assumptions we are making right now and adjusting when necessary. (Of course unless there is already something like that in the pipeline and we know it will break soon).

Reply:
Well there's https://github.com/scala/scala/pull/8509 in the pipeline, which, IIUC, will generate public static methods for default getters. Default getters have mangled names, so they shouldn't collide with static forwarders.

Reply:
Although that PR upstream is not going to be relevant any time soon, since it breaks binary compat and therefore would only go to 2.14.x, not 2.13.x.

Reply:
> Since this is not necessary right now, personally, I'd feel more comfortable with checking the (more rigid) assumptions we are making right now and adjusting when necessary.

Done now.